### PR TITLE
misc cleanups re sp-math-all, FIPS, smallstack

### PR DESCRIPTION
**Pull request by douzzer**
Main attractions here are refactoring `_word()` wrapper functions in src/ssl.c to work with `WOLFSSL_BN_ULONG` defined as target-native `unsigned long` (rather than `mp_digit`), and a slew more `_SMALL_STACK` refactors to keep `--enable-all --enable-sp-math-all --enable-smallstack` `wolfcrypt_test()` under 16k of stack.  `--enable-linuxkm[-defaults]` now uses sp-math-all by default, and keeps `wolfcrypt_test()` under 10k of stack.

### CODE REVIEW

File: wolfssl/openssl/bn.h [hunk 1]
```diff
@@ -85,7 +85,7 @@ WOLFSSL_API int wolfSSL_BN_is_zero(const WOLFSSL_BIGNUM*);
 WOLFSSL_API int wolfSSL_BN_is_one(const WOLFSSL_BIGNUM*);
 WOLFSSL_API int wolfSSL_BN_is_odd(const WOLFSSL_BIGNUM*);
 WOLFSSL_API int wolfSSL_BN_is_negative(const WOLFSSL_BIGNUM*);
-WOLFSSL_API int wolfSSL_BN_is_word(const WOLFSSL_BIGNUM*, WOLFSSL_BN_ULONG);
+WOLFSSL_API int wolfSSL_BN_is_word(const WOLFSSL_BIGNUM*, unsigned long);
```
Comment (dgarske) [inline wolfssl/openssl/bn.h:88]: @douzzer , why not just fix `WOLFSSL_BN_ULONG` to be `unsigned long`? Also unsigned long will be different sizes depending on the architecture. Example: 16-bit micro unsigned long is 32-bit most others is 64-bit.

File: src/ssl.c [hunk 2]
```diff
@@ -47012,9 +47052,30 @@ WOLFSSL_BN_ULONG wolfSSL_BN_mod_word(const WOLFSSL_BIGNUM *bn,
         return (WOLFSSL_BN_ULONG)WOLFSSL_FATAL_ERROR;
     }
 
-    if (mp_mod_d((mp_int*)bn->internal, w, &ret) != MP_OKAY) {
-        WOLFSSL_MSG("mp_add_d error");
-        return (WOLFSSL_BN_ULONG)WOLFSSL_FATAL_ERROR;
+    if (w <= MP_MASK) {
+        WOLFSSL_BN_ULONG bn_ret;
```
Comment (dgarske) [inline src/ssl.c:47056]: Why not use `mp_digit` here and fix uses of `WOLFSSL_BN_ULONG` to define `unsigned long`?

**Review by dgarske — CHANGES_REQUESTED** [review]

File: wolfssl/openssl/bn.h (hunk 1 above)
Comment (douzzer) [reply wolfssl/openssl/bn.h:88]: Oh hey good idea, how 'bout that, `WOLFSSL_BN_ULONG` is only used in ssl.c/bn.h.  It had the sound of something used internally so I hadn't considered changing its meaning.  A quick survey suggests changing it from `mp_digit` to `unsigned long` would have exactly the result I'd intended.

Btw, notice that `wolfSSL_BN_get_word()` was already returning a target-native `unsigned long`.  My patch is just intended to harmonize the API with that.  The substantive change, of course, is that all of these routines now know how to convert multiple mp/sp digits to a single target-native unsigned long when it will fit, which is to say, they DTRT when `[ms]p_digit` is smaller than `unsigned long`, and only error when the MP/SP number can't fit in the native long word.

Re your second point, the purpose of the patch is indeed to arrange for these API wrapper functions to follow the target-native `unsigned long` word size, to meet caller expectations of OpenSSL semantics, and in particular, to not truncate 65537 with a coercion to 16 bit `unsigned short` in the 32 bit SP use case.

I'll push code shortly that restores the cosmetics of the bn.h API, while defining `WOLFSSL_BN_ULONG` as target-native `unsigned long` rather than `mp_digit`/`sp_digit`.

File: src/ssl.c (hunk 2 above)
Comment (douzzer) [reply src/ssl.c:47056]: Indeed.  See above.

**Review by dgarske — CHANGES_REQUESTED** [review]
Hey @douzzer , great work. Couple things.

1. Need cast on XMALLOC here:

```
./configure CC=g++ --enable-all --enable-sp-math-all --enable-smallstack --enable-smallstackcache && make
...
wolfcrypt/src/asn.c:1196:18: error: assigning to 'IndefItems *' from incompatible type 'void *'
    indefItems = XMALLOC(sizeof(IndefItems), NULL, DYNAMIC_TYPE_TMP_BUFFER);
                 ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

2. Some of the `const char*` in test.c were changed to `static const char*` and that is less than ideal for embedded systems.

David

**Comment by douzzer** [comment]
@dgarske fixed the `XMALLOC()` cast issue in asn.c (Sean's blame) and two more besides, one in chacha20_poly1305.c (my blame) and one in wc_pkcs11.c (Sean's), plus added explicit `(void *)` casts to half a dozen previously uncasted calls, to make my simple `find | xargs pcre2grep` recipe happy that the codebase is clean of the problem now.

As for the `static const` test vectors in test.c, AFAIK those changes are purely reductive.  The vectors have to exist in the text segment either way, and making them `static` just keeps them off the stack.  I've made that change at innumerable other test.c 
codepoints and it should be close to complete now:

```
$ fgrep -c 'static const' wolfcrypt/test/test.c 
526
```

**Review by ejohnstown — APPROVED** [review]

**Review by dgarske — APPROVED** [review]

**Review by SparkiDev — APPROVED** [review]

File: wolfssl/openssl/aes.h [hunk 3]
```diff
@@ -42,32 +42,7 @@
...
-    #ifdef WOLFSSL_AFALG
-    void* afalg_holder[288 / sizeof(void*)];
-    #endif
-    #ifdef HAVE_PKCS11
-    void* pkcs11_holder[(AES_MAX_ID_LEN + sizeof(int)) / sizeof(void*)];
-    #endif
-    #if defined(WOLFSSL_ASYNC_CRYPT) || defined(WOLF_CRYPTO_CB)
-    void* async_holder[128 / sizeof(void*)];
-    #endif
+    void *buf[(sizeof(Aes) / sizeof(void *)) + 1];
```
Comment (toddouska) [inline wolfssl/openssl/aes.h:45]: 16byte alignment was required on some platforms with some builds.  "void*" is not guaranteed to have 16byte alignment.  Why was this removed?

Comment (douzzer) [reply wolfssl/openssl/aes.h:45]: Refactored because `async_holder` was smaller than `Aes` with some configs, i.e. buffer overrun bug.

The new code has identical alignment dynamics to the old code as far as I can see.  Array of `(void *)` before and after.

Comment (douzzer) [reply wolfssl/openssl/aes.h:45]: OIC, the github-generated excerpt above doesn't show the original code.  `ALIGN16` is indeed in the original and I'll restore it and get that pushed momentarily.

Thanks for catching this.

**Review by toddouska — APPROVED** [review]
